### PR TITLE
AutoNLP using AutoVIML library for Natural Language Processing

### LINKED ISSUE #6038: [Machine Learning] AutoNLP using AutoVIML library for Natural Language Processing
#### Proposed title of article
[Machine Learning] AutoNLP using AutoVIML library for Natural Language Processing

#### Proposed article introduction
Natural language processing is the branch of computer science that enables machines to understand the text. This process is usually hectic and has many processes. But with the use of AutoVIML which is an AutoNLP library natural language processing has been made easier. AutoVIML builds more robust models that are more effective to perform NLP tasks.

AutoVIML automates the processes of EDA and text processing to format the dataset correctly. It also automates the processes of stemming, tokenization, lemmatization, CounterVectorizer, and TfidfVectorizer. It automates this process using a machine learning Pipeline, this saves the developers time and makes the process of natural language processing easier.

In this tutorial, we build a product review model using AutoVIML. AutoVIML will be able to build a model that classifies customers reviews as either positive or negative.

#### Key takeaways 
1. What is AutoNLP?
2. Installing AutoVIML.
3. Dataset preparation using AutoVIML.
4. Creating a machine learning pipeline to automate the NLP processes.
5. Building the product review model using AutoVIML.

#### Article quality
In this tutorial, we will first explain the stages involved in natural language processing in detail. This will give a reader a deeper understanding of NLP concepts before introducing the AutoNLP concepts. This tutorial will also cover the important  AutoVIML functions that help with the automation process. This tutorial will be detailed and have guided steps that a reader can easily follow. We will finally build a custom product review model.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by jamesomina99**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6038

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by srishilesh** [comment]
Hello @jamesomina99 

Thanks for submitting this PR.

Your PR has entered our reviewing queue. Please wait until a peer-reviewer pings you with the changes to be made.

Thanks.

**Comment by wilkistermumbi** [comment]
@ahmadmardeni1 @WanjaMIKE @LinusMuema I'd like to review this PR.

cc: @hectorkambow

**Comment by WanjaMIKE** [comment]
> @ahmadmardeni1 @WanjaMIKE @LinusMuema I'd like to review this PR.
> 
> cc: @hectorkambow

@wilkistermumbi I think you were added to the peer review team. You are free to take up PRs that you can review. Thanks

**Comment by wilkistermumbi** [comment]
> > @ahmadmardeni1 @WanjaMIKE @LinusMuema I'd like to review this PR.
> > cc: @hectorkambow
> 
> @wilkistermumbi I think you were added to the peer review team. You are free to take up PRs that you can review. Thanks

@WanjaMIKE I still haven't been granted reviewer access to the Eng-Ed repo. You can help me add the necessary labels in the mean time.

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by ahmadmardeni1** [comment]
Let me check this out. @wilkistermumbi You will get access soon.

**Comment by ahmadmardeni1** [comment]
You have been added to the "peer reviewers" team. Please check again, you can edit the labels. @Wilkistermumbi

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+- `helpful_votes`
+It is the number of votes of a purchased product.
+
+- `review_headline`
+It is the title product review.
+
+- `review_body`
+It is a detailed description of the review.
+
+After loading the dataset, we need to convert the dataset into an array using NumPy. An array can be easily be used by the model.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:128]: Please rewrite the second sentence in this paragraph in active voice.

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+To see the structure of the new dataset, use this code:
+
+```python
+dataset
+```
+The output is shown below:
+
+![Dataset array](/engineering-education/autonlp-using-autoviml-library-for-natural-language-processing/dataset-array.jpg)
+
+From the image above our dataset is in array format. The next step is to extract the four important columns from our dataset.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:150]:
```suggestion
From the image above, our dataset is in array format. The next step is to extract the four important columns from our dataset.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+```python
+helpful_votes=dataset['data']['helpful_votes']
+review_headline=dataset['data']['review_headline']
+review_body=dataset['data']['review_body']
+rating=dataset['data']['star_rating']
+```
+This will enable us to use these four columns as inputs for our model during training. Next, we will create a data frame for our dataset. A data frame will neatly organize our dataset into columns and rows.
+
+### Creating a data frame
+To create a data frame we need the Pandas library. Let's import the Pandas library.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:164]:
```suggestion
To create a data frame, we need the Pandas library. Let's import the Pandas library.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+reviews_df
+```
+The output is shown below:
+
+![Dataframe structure](/engineering-education/autonlp-using-autoviml-library-for-natural-language-processing/dataframe-structure.jpg)
+
+From the image above, our dataset is neatly organized into rows and columns. It has four columns and 85981 rows.
+
+### Adding the target column
+We need to add a target column. The target column represents the model output after making a prediction. The model classifies customers reviews as either positive or negative. Positive reviews are represented by 1 while negative reviews are represented by 0.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:203]:
```suggestion
We need to add a target column. The target column represents the model output after making a prediction. The model classifies customers reviews as either positive or negative. Positive reviews are represented by `1`, while negative reviews are represented by `0`.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+```python
+reviews_df
+```
+The output is shown below:
+
+![Target column](/engineering-education/autonlp-using-autoviml-library-for-natural-language-processing/target-column.jpg)
+
+The image above shows the dataset with the added target column. The target column is either labeled `1` or `0`.
+
+Our dataset is now neatly and correctly formatted, we now need to split our dataset into two sets.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:223]:
```suggestion
Our dataset is neatly and correctly formatted. We now need to split our dataset into two sets.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+### Dataset splitting
+We will split our dataset into two sets. One set will be used to train the model, the other set used to test the model.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:226]:
```suggestion
We will split our dataset into two sets. We use one set to train the model, while the other set to test the model. 
```

Comment (wilkistermumbi) [reply content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:226]: Avoid using passive voices in your sentences

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+```python
+from sklearn.model_selection import train_test_split
+```
+To split the dataset, use this code:
+
+```python
+train, test = train_test_split(reviews_df, test_size=0.25)
+```
+From the code above we have used `test_size=0.25`. This is the ratio used for dataset splitting. 75% of the data is used for training and 25% is used for testing.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:238]:
```suggestion
From the code above, we have used `test_size=0.25`. It is the ratio used for dataset splitting. We use 75% of the data for training and 25% for testing.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+After splitting the dataset, let's now use AutoVIML to automate natural language processing.
+
+### Using AutoVIML
+AutoVIML has in-built functions that are used to automate natural language processing. We will use the `Auto_NLP` function which is imported from AutoVIML.
+
+To import `Auto_NLP` use this code:
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:245]:
```suggestion
To import `Auto_NLP`, use this code:
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+It is the output column of the model after making a prediction.
+
+- `train`
+It is the split dataset used for training.
+
+- `test`
+It is the test dataset used for testing.
+
+- `score_type='balanced_accuracy'`
+It is used to calculate the accuracy score for the model.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:287]:
```suggestion
It calculates the accuracy score for the model.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+- `modeltype='Classification'`
+It specifies the type of model we are building. We are building a model used for classification.
+
+- `top_num_features=50`
+It specifies the number of important features used during training. Features are the important attribute found in the dataset.
+
+- `build_model=True`
+It is used to tell the `Auto_NLP` function to build the model. `Auto_NLP` function will use the key AutoVIML features and produce an optimized model.
+
+After running this code, the process will run produce outputs. The outputs shows the model building progress. These outputs will also show us which parameters and algorithms have been used to build the model.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:298]:
```suggestion
After running this code, the process will produce outputs. The outputs show the model-building progress. These outputs will also show us which parameters and algorithms have been used to build the model.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+The fourth output is shown below:
+
+![Output 4](/engineering-education/autonlp-using-autoviml-library-for-natural-language-processing/punctuation-hashtags.jpg)
+
+This output shows the number of punctuations and hashtags in our dataset.
+
+The next output shows the vectorization process.
+
+### Vectorization process
+The vectorization process converts the text data into numeric data which the model can use. Machine learning models do not understand the raw text.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:331]:
```suggestion
The vectorization process converts the text data into numeric data, which the model can use. Machine learning models do not understand the raw text.
```

File: content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+From the image above, the best vectorizer selected is the TFIDF vectorizer. The `Auto_NLP` function also adds the best parameters for the TFIDF vectorizer.
+
+The next output shows the best algorithm and the machine learning pipeline.
+
+### The best algorithm and machine learning pipeline
+The `Auto_NLP` function searches through all of the available algorithms. It then selects the one with the best results. The selected algorithm is shown in the output below.
+
+![Best algorithm](/engineering-education/autonlp-using-autoviml-library-for-natural-language-processing/best-algorithm.png)
+
+From the image above the best algorithm is the `Multinomial Naive Bayes algorithm`.
```
Comment (wilkistermumbi) [inline content/articles/autonlp-using-autoviml-library-for-natural-language-processing/index.md:350]: Add commas where necessary

**Comment by wilkistermumbi** [comment]
@jamesomina99 This is a nice tutorial. However, the minor grammatical errors such as lack of proper comma use and uses of passive voices mask the beauty of the tutorial.In addition to my revisions, go through the whole tutorial and fix mistakes I might have missed so as to avoid further revisions. Do not over-rely on the basic grammarly. Unless you've subscribed for the premium version of grammarly, always proofread your work to identify mistakes which grammarly might have failed to detect.

**Comment by jamesomina99** [comment]
@wilkistermumbi, I have gone through the entire article and made the necessary changes. I have also committed the suggested changes. Kindly review.

**Comment by hectorkambow** [comment]
👍

**Comment by hectorkambow** [comment]
@wilkistermumbi 
If anymore edits are needed lets go ahead and close the PR

**Comment by hectorkambow** [comment]
@wilkistermumbi  Any update on this?

**Comment by wilkistermumbi** [comment]
> @wilkistermumbi Any update on this?

@hectorkambow It's ready for final review.

**Comment by lalith1403** [comment]
Jumping on this one, team 🚀

**Comment by jamesomina99** [comment]
Hello @lalith1403, what is the status of this PR?

**Comment by hectorkambow** [comment]
> Hello @lalith1403, what is the status of this PR?

I'll be doing the final review on this one. 
No worries

**Comment by hectorkambow** [comment]
@jamesomina99
